source/MRMesh/MROpenVDBHelper.h

```diff
@@ -0,0 +1,213 @@
+ * @details similar to openvdb::RangeProcessor
+ * @tparam TreeT tree type
+ * @tparam Transformer functor to transform tree
+ */
+template <class TreeT, typename Transformer>
+class RangeProcessor
+{
+public:
+    using InterruptFunc = std::function<bool( void )>;
+
+    using ValueT = typename TreeT::ValueType;
+
+    using LeafIterT = typename TreeT::LeafCIter;
+    using TileIterT = typename TreeT::ValueAllCIter;
+    using LeafRange = typename openvdb::tree::IteratorRange<LeafIterT>;
+    using TileRange = typename openvdb::tree::IteratorRange<TileIterT>;
+
+    using InTreeAccessor = typename openvdb::tree::ValueAccessor<const TreeT>;
+    using OutTreeAccessor = typename openvdb::tree::ValueAccessor<TreeT>;
+
+    RangeProcessor( const openvdb::math::CoordBBox& b, const TreeT& inT, TreeT& outT, const Transformer& xform ) :
+        mIsRoot( true ), mXform( xform ), mBBox( b ),
+        mInTree( inT ), mOutTree( &outT ), mInAcc( mInTree ), mOutAcc( *mOutTree )
+    {}
+
+    RangeProcessor( const openvdb::math::CoordBBox& b, const TreeT& inTree, const Transformer& xform ) :
+        mIsRoot( false ), mXform( xform ), mBBox( b ),
+        mInTree( inTree ), mOutTree( new TreeT( inTree.background() ) ),
+        mInAcc( mInTree ), mOutAcc( *mOutTree )
+    {}
+
+    ~RangeProcessor()
+    {
+        if ( !mIsRoot ) delete mOutTree;
+    }
+
+    /// Splitting constructor: don't copy the original processor's output tree
+    RangeProcessor( RangeProcessor& other, tbb::split ) :
+        mIsRoot( false ),
+        mXform( other.mXform ),
+        mBBox( other.mBBox ),
+        mInTree( other.mInTree ),
+        mOutTree( new TreeT( mInTree.background() ) ),
+        mInAcc( mInTree ),
+        mOutAcc( *mOutTree ),
+        mInterrupt( other.mInterrupt )
+    {}
+
+    void setInterrupt( const InterruptFunc& f ) { mInterrupt = f; }
+
+    /// Transform each leaf node in the given range.
+    void operator()( const LeafRange& rCRef )
+    {
+        LeafRange r = rCRef;
+        for ( ; r; ++r )
+        {
+            if ( interrupt() ) break;
+
+            LeafIterT i = r.iterator();
+            openvdb::math::CoordBBox bbox( i->origin(), i->origin() + openvdb::math::Coord( i->dim() ) );
+            if ( !mBBox.empty() )
+            {
+                // Intersect the leaf node's bounding box with mBBox.
+                bbox = openvdb::math::CoordBBox(
+                    openvdb::math::Coord::maxComponent( bbox.min(), mBBox.min() ),
+                    openvdb::math::Coord::minComponent( bbox.max(), mBBox.max() ) );
+            }
+            if ( !bbox.empty() )
+            {
+                openvdb::math::Coord bboxMin = bbox.min();
+                openvdb::math::Coord bboxMax = bbox.max();
+                for ( int iz = bboxMin.z(); iz < bboxMax.z(); ++iz )
+                for ( int iy = bboxMin.y(); iy < bboxMax.y(); ++iy )
+                for ( int ix = bboxMin.x(); ix < bboxMax.x(); ++ix )
```

Comment:
Mb use bbox.begin() bbox.end() (to have one cycle)

Reply:
reworked